grove/grove_co2_scd30.py

```diff
@@ -0,0 +1,172 @@
+#
+
+'''
+## License
+
+The MIT License (MIT)
+
+Copyright (C) 2022  matsujirushi
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+'''
+
+from typing import NoReturn
+from grove.i2c import Bus
+import struct
+import time
+
+
+class GroveCo2Scd30(object):
+
+    def __init__(self, address=0x61, bus=None):
+        self.address = address
+        self.bus = Bus(bus)
+
+        self.set_measurement_interval(2)
+        self.trigger_continuous_measurement()
+
+    @staticmethod
+    def _calc_crc(data: list) -> int:
+        crc = 0xff
+
+        for d in data:
+            crc ^= d
+
+            for _ in range(8):
+                if crc & 0x80:
+                    crc = ((crc << 1) ^ 0x31) & 0xff
+                else:
+                    crc = (crc << 1) & 0xff
+
+        return crc
+
+    def _write(self, cmd: int, data: list):
+        write_data = list(struct.pack(">H", cmd))
+        if data is not None:
+            for d in data:
+                write_data.extend(struct.pack(">H", d))
+                write_data.append(GroveCo2Scd30._calc_crc(struct.pack(">H", d)))
+
+        write_msg = self.bus.msg.write(self.address, write_data)
+        self.bus.i2c_rdwr(write_msg)
+
+    def _read(self, address: int, data_number: int) -> list:
+        write_data = list(struct.pack(">H", address))
+
+        write_msg = self.bus.msg.write(self.address, write_data)
+        self.bus.i2c_rdwr(write_msg)
+
+        time.sleep(0.003)
+
+        read_msg = self.bus.msg.read(self.address, 3 * data_number)
+        self.bus.i2c_rdwr(read_msg)
+
+        result = []
+        for i in range(data_number):
+            d = read_msg.buf[i*3:i*3+2]
+            if GroveCo2Scd30._calc_crc(d) != read_msg.buf[i*3+2][0]:
+                raise ValueError("CRC mismatch")
+
+            result.append(struct.unpack(">H", d)[0])
+
+        return result
+
+    def trigger_continuous_measurement(self, pressure: int = 0):
+        self._write(0x0010, [pressure])
```

Comment:
I think it's better to add macro definitions at the beginning of the file with all the I2C command or register.
Because it make the code easier to understan and more compatible.

Reply:
Added constants.